# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/manager/ssl/tests/unit/head_psm.js
# Commit: 16350e08d34f
# Full Hash: 16350e08d34ff06c7f46745c6b160c7a7bda9140
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2019-12-06 04:43:57
# Regressor Bug: 1586855
# File Overlap Count: 6
# Description:
#   bug 1586855 - incorporate CRLite filters into cert_storage r=jcj,kjacobs
#   
#   This patch implements CRLite lookups for TLS server certificate revocation
#   information in telemetry-only mode. It adds a new preference
#   "security.pki.crlite_mode" to control the behavior of this feature. Setting
# ==============================================================================

diff -r 64fd40663930 -r 16350e08d34f security/manager/ssl/tests/unit/head_psm.js
--- a/security/manager/ssl/tests/unit/head_psm.js	Thu Dec 05 22:23:54 2019 +0000
+++ b/security/manager/ssl/tests/unit/head_psm.js	Thu Dec 05 22:41:53 2019 +0000
@@ -22,6 +22,8 @@
   "resource://gre/modules/XPCOMUtils.jsm"
 );
 
+const { X509 } = ChromeUtils.import("resource://gre/modules/psm/X509.jsm");
+
 const isDebugBuild = Cc["@mozilla.org/xpcom/debug;1"].getService(Ci.nsIDebug2)
   .isDebugBuild;
 
@@ -116,6 +118,10 @@
 
 const NO_FLAGS = 0;
 
+const CRLiteModeDisabledPrefValue = 0;
+const CRLiteModeTelemetryOnlyPrefValue = 1;
+const CRLiteModeEnforcePrefValue = 2;
+
 // Convert a string to an array of bytes consisting of the char code at each
 // index.
 function stringToArray(s) {
@@ -262,7 +268,8 @@
   usage,
   time,
   /* optional */ isEVExpected,
-  /* optional */ hostname
+  /* optional */ hostname,
+  /* optional */ flags = NO_FLAGS
 ) {
   return new Promise((resolve, reject) => {
     let result = new CertVerificationExpectedErrorResult(
@@ -271,7 +278,7 @@
       isEVExpected,
       resolve
     );
-    certdb.asyncVerifyCertAtTime(cert, usage, NO_FLAGS, hostname, time, result);
+    certdb.asyncVerifyCertAtTime(cert, usage, flags, hostname, time, result);
   });
 }
 
@@ -1163,3 +1170,15 @@
 
   return testModule;
 }
+
+// Given an nsIX509Cert, return the bytes of its subject DN (as a JS string) and
+// the sha-256 hash of its subject public key info, base64-encoded.
+function getSubjectAndSPKIHash(nsCert) {
+  let certBytes = nsCert.getRawDER();
+  let cert = new X509.Certificate();
+  cert.parse(certBytes);
+  let subject = cert.tbsCertificate.subject._der._bytes;
+  let subjectString = arrayToString(subject);
+  let spkiHashString = nsCert.sha256SubjectPublicKeyInfoDigest;
+  return { subjectString, spkiHashString };
+}